Repro for #12651

nemanjaglumac opened the pull request:
### What does this PR accomplish?
- reproduces the issue #12651 
- checks for the link structure when converting a question to SQL/Native (it should remove `/notebook`)

### Where should the reviewer start?
1. `yarn test-cypress-open`
2. click on `/scenarios/question/new.cy.specs.js`

### How should this be manually tested?
- steps to reproduce from #12651 can be used for manual testing as well

### Screenshots
N/A

### Notes/Questions
Any other method to get to the open orders table other than
```javascript
    cy.server();
    cy.route("POST", "/api/dataset").as("dataset");
    openOrdersTable();
    cy.wait("@dataset");
```
was resulting in `(XHR) POST (aborted) /api/dataset`.
- test should pass after the underlying issue is solved

nemanjaglumac committed:
Add repro for #12651

salsakran commented:
@paulrosenzweig  @flamber can you take a look at this PR ?

flamber reviewed (commented):
All the other tests seems to use text to find the buttons, why is this using `data-metabase-event` attribute?

nemanjaglumac commented:
> All the other tests seems to use text to find the buttons, why is this using `data-metabase-event` attribute?

@flamber - it was my understanding that searching/getting DOM elements by their css classes was discouraged. Since those two buttons don't have any text (only icons), I thought that targetting their "data" attribute was somehow the safer way.

Now that you pointed it out, it does seem maybe overly complicated compared to the other tests. I can easily switch to `cy.get(".Icon-notebook").click();` if you don't mind the css class?

flamber commented:
@nemanjaglumac Right, didn't think of that, just looked at the other notebook tests. I don't know what the preferred way is.

paulrosenzweig commented on frontend/test/metabase/scenarios/question/new.cy.spec.js:
```diff
@@ -25,6 +25,19 @@ describe("scenarios > question > new", () => {
       cy.contains("37.65");
     });
+
+    it("should remove `/notebook` from URL when converting question to SQL/Native", () => {
```
This should be `it.skip` so we can merge the test in before the issue is fixed. Also, I think it'd be good to include the issue number, so we can find the repro test easily. Here's [an example](https://github.com/metabase/metabase/pull/12914/files#diff-776068597c6aff5c2add7a9b3615d32aR24).

nemanjaglumac replied:
Aaaaah, ok. Now it makes sense. I believe @Opalevanescence mentioned `.skip` in one of her comments, but I obviously didn't connect the dots :). Also, @paulrosenzweig thank you for the example.

I'll fix this and push new changes soon.

nemanjaglumac commented:
> @nemanjaglumac thanks for all the PRs!

You're welcome. Thank YOU for putting the time to explain this better to me. It helps :)

nemanjaglumac committed:
Refactor test

- stick to CSS classes when targetting icons

nemanjaglumac committed:
Skip the test

- skip the test until the issue is resolved
- added reference to the original issue in the test description

nemanjaglumac commented on frontend/test/metabase/scenarios/question/new.cy.spec.js:
```diff
@@ -32,9 +32,10 @@ describe("scenarios > question > new", () => {
       cy.wait("@dataset");
       cy.url().should("include", "question#");
-      cy.get("[data-metabase-event$='Go to Notebook Mode']").click();
+      // Isolate icons within "QueryBuilder" scope because there is also `.Icon-sql` in top navigation
```
@paulrosenzweig - this is why targetting CSS is tricky. However, I believe that scoping those icons within "QueryBuilder" should do the job.

paulrosenzweig replied:
Yeah... typically icons are unique on a page, but unfortunately not here. I like using `.QueryBuilder` for scoping. If we end up using that a lot we could add a helper `queryBuilder()` like we have for `modal()`, `sidebar()`, etc.

Opalevanescence reviewed (approved):
Looks good to me, if you're ready to merge.

nemanjaglumac commented:
@paulrosenzweig - do you want me to rebase (and do you encourage rebase) in order to resolve merge conflict?

What's the common practice that you guys prefer?

Note: this happened because the other PR introduced changes to the same file as this one. In retrospect, maybe I could've put them together 😔.

flamber commented:
@nemanjaglumac Yes, please resolve the conflict. I think GitHub recommends merge, but rebase should be fine too.

nemanjaglumac commented:
> @nemanjaglumac Yes, please resolve the conflict. I think GitHub recommends merge, but rebase should be fine too.

@flamber - I will have to `git push --force` and rewrite history. Since nobody else was working on this branch, my guess is that you'd be ok with it? I have to ask nevertheless. Force pushing should always be approached with care.

It is my understanding that only someone with write access could do the merge, since one has to be on the master branch to merge the remote. If I got something wrong, please tell me.

P.S. I tested rebase locally. Everything's fine. The only thing that I'm unsure of is wether it will remove @paulrosenzweig 's review and our comments on the code?

flamber commented:
@nemanjaglumac Try. The worst thing that can happen is you just close this PR and open a new if it gets borked.
Otherwise merge, like the built-in function in GitHub: https://github.com/metabase/metabase/pull/12919/conflicts

paulrosenzweig commented:
> since one has to be on the master branch to merge the remote. If I got something wrong, please tell me.

I think you should be able to merge `metabase:master` into `nemanjaglumac:repro/issue_12651` locally. You'll have to fix the conflicts as part of that merge commit on `nemanjaglumac:repro/issue_12651`. You can then push that here and it will merge cleanly back into `metabase:master` in the Github UI.

But I think a rebase and force push is fine too!

> The only thing that I'm unsure of is wether it will remove @paulrosenzweig 's review and our comments on the code?

It will probably break the comments but not the approval.

robdaemon commented:
@nemanjaglumac force pushing to a branch you're working on is just fine - the rest of us do it all the time because of merge conflicts. It's especially safe since you're working on a fork.

Here's how I resolve conflicts so a PR can be merged:

```bash
git fetch upstream
git rebase upstream/master
git mergetool
# resolve all the conflicts in the merge, then rerun the tests
git rebase --continue
# loop through rebasing until all conflicts are resolved
git push -f origin MyBranchName
```

(assuming you've done: `git remote add upstream https://github.com/metabase/metabase.git`

robdaemon commented:
Of note, any comment on the code doesn't really get removed, it gets marked as outdated, so it's totally okay.

robdaemon commented:
Ohhhh one more thing - rebase is better for this case - merge can have unintended side-effects, potentially losing a change you've made when there's a conflict in the same file that's newer than yours, and it introduces an additional merge commit.

Rebase attempts to "replay" your changes, slotted in to the commit history to line up with the other commits based on time/date. You likely see more "conflicts" that way that you need to resolve, but it's easier to make sure you don't lose anything.

nemanjaglumac commented:
@flamber - good point about GH merge. Tbh, I was always "wary" of doing anything related to Git outside of terminal, but this was really minor change so your suggestion was actually the easiest.

@paulrosenzweig - yup. I was just thinking the other way around (merging remote into master).

@robdaemon - thank you. That was exactly my workflow and what I did on local branches. And also thanks for clarifying the part about the comments.

In general, thank you all for valuable input.

nemanjaglumac commented:
Ahhhhh, lint test failed. (one more reason to do everything locally :))

Back to the "drawing board". Git rebase it is.

robdaemon commented:
![Screenshot_20200716_142419](https://user-images.githubusercontent.com/287213/87833598-18a13600-c83d-11ea-94de-bb5af921d39b.png)

This was me yesterday because I kept botching small things that would break the linters for the Clojure code :grin:

nemanjaglumac commented:
How is it possible that CircleCi fails tests that haven't been the subject of this PR? Even locally some of them are failing when `yarn test`. Do we ignore them because they are being ported to Cypress tests?

paulrosenzweig commented:
@nemanjaglumac that failure was almost definitely a flake. I just triggered a rerun on that test.

There's some flakiness both in the legacy e2e tests and Cypress, but it's often still worth considering whether it could be a real failure. Earlier this week I reran a failed e2e test twice before I believed it was real and looked into it 😬

robdaemon commented:
Yep, probably a flake, but since we run the entire battery of tests with every PR, you're going to see that occasionally, or a failure if the base branch you used already had a failure in it.

nemanjaglumac commented:
Thank you both for really quick answers (on Friday! 🤓).
